lib/linked-list.js

```diff
@@ -125,42 +214,68 @@ class LinkedList {
     /*
     method to reverse the singly linked list
     note: the nodes should be moved and not just the values in the nodes
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     reverse() {
-        throw new Error("This method hasn't been implemented yet!");
+        if (!this.#head) return;
+
+        let previous = null;
+        let current = this.#head;
+        let next = null;
+
+        while (current) {
+            next = current.next;
+            current.next = previous;
+            previous = current;
+            current = next;
+        }
+        this.#head = previous;
     }
 
     // Advanced Exercises
     /*
     returns the value at the middle element in the singly linked list
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n) (because of getAtIndex)
+    Space Complexity: O(1)
      */
     findMiddleValue() {
-        throw new Error("This method hasn't been implemented yet!");
+        let middleIndex = Math.floor(this.#listLength/2);
+        return this.getAtIndex(middleIndex);      
     }
 
     /*
     find the nth node from the end and return its value
     assume indexing starts at 0 while counting to n
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n) (because of getAtIndex)
+    Space Complexity: O(1)
     */
     findNthFromEnd(n) {
-        throw new Error("This method hasn't been implemented yet!");
+        let nthIndex = this.#listLength - n;
+        return this.getAtIndex(nthIndex);
     }
 
     /*
     checks if the linked list has a cycle. A cycle exists if any node in the
     linked list links to a node already visited.
     returns true if a cycle is found, false otherwise.
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     hasCycle() {
```

Comment:
Ah so that's the reason for the flag.  Note you could also use a Set, or use a fast and slow reference.